timely_beliefs/beliefs/classes.py

```diff
@@ -1467,6 +1492,124 @@ def resample_events(
+
+        # Check assumptions
+        if self.sensor.knowledge_time(event_time_window[0]) < belief_time:
+            # No backcasting
+            raise NotImplementedError("Backcasting is not implemented.")
+        if self.empty:
+            # At least 1 belief
+            raise NotImplementedError(
+                "Cannot form new beliefs given an empty data frame."
+            )
+        if self.lineage.number_of_sources != 1:
+            # Single source only
+            raise NotImplementedError(
+                "Cannot form new beliefs given multi-sourced data. Consider picking 1 source, e.g. using: df = df[df.index.get_level_values('source') == df.lineage.sources[0]]"
+            )
+        if self.lineage.probabilistic_depth != 1:
+            # Deterministic beliefs only
+            raise NotImplementedError(
+                "Cannot form new beliefs given probabilistic data. Consider pre-computing deterministic beliefs, e.g. using: df = df.make_deterministic()"
+            )
+        if self.lineage.number_of_beliefs != self.lineage.number_of_events:
+            # One belief per event
+            raise NotImplementedError(
+                "Cannot form new beliefs given multiple beliefs about the same event. Consider picking a single belief per event, e.g. using: df = belief_utils.select_most_recent_belief(df)"
+            )
+
+        df = self
+
+        # Do NOT use future data to predict NOR to fit
+        if "belief_time" not in df.index.names:
+            df = df.convert_index_from_belief_horizon_to_time()
+        df = df[df.index.get_level_values("belief_time") < belief_time]
+
+        # Simplify index
+        df = df.reset_index()[["event_start", "event_value"]].set_index("event_start")
+
+        # Convert to local time in UTC, which sktime expects
+        tz = event_time_window[0].tzinfo
+        utc = pytz.utc
+        event_time_window = tuple(
+            dt.astimezone(utc).replace(tzinfo=None) for dt in event_time_window
+        )
+        df.index = df.index.tz_convert(utc).tz_localize(None)
+
+        # Resample to the resolution the data already has, just to set the index frequency, which sktime expects
+        df = df.reindex(
+            pd.date_range(
+                df.index[0],
+                event_time_window[0],
+                freq=df.event_resolution,
+                closed="left",
+            )
+        )
+        df = df.resample(df.event_resolution).mean()
+
+        # todo: if forecaster does not handle missing values, impute intermediate missing values and forecast trailing missing values
+
+        # Apply model
+        if isinstance(forecaster, BaseForecaster):
+            forecast_event_starts = pd.date_range(
+                event_time_window[0],
+                event_time_window[1],
+                freq=df.event_resolution,
+                closed="left",
+            )
+            forecaster.fit(df["event_value"])
```

Comment:
At this time, our BeliefDataFrames do not support regressors to be used here, correct? For example, forecasting wind speed based on time of day and temperature - how would we do that? (maybe another ticket)

Reply:
Good idea to separate this into another ticket. It can be done by passing an extra `X` to the fit function, for example, like this:
```
X=df.loc[:, df.columns != "event_value"]
```
Here, I'm assuming the external regressors are given as additional columns in the BeliefsDataFrame.

Testing it with sktime's naive forecaster will just lead to the forecaster ignoring `X` though, so that doesn't feel like a wholesome test.

Reply:
Did we make an issue out of this yet? I can't find one.

Reply:
Ah, I see you already add the `X` in this PR. I believe a little comment would help.

And is there a case where columns in X would be bad to include?

We could also add ab optional parameter to `form_beliefs`, e.g. "X_column_names".

Reply:
I created https://github.com/SeitaBV/timely-beliefs/issues/116 just now.